_posts/2014-11-13-cockpit-plugin-tutorial.md

```diff
@@ -78,92 +77,101 @@ Lets take a look at the pinger HTML, and see how it works.
+
+```js
 function ping_run() {
+}
+...
+button.addEventListener("click", ping_run);
 ```
 
-In the `ping_run()` function is where the magic happens. `cockpit.spawn` is a function, [documented here](https://cockpit-project.org/guide/latest/api-cockpit.html#latest-spawn) that lets you spawn processes on the server and interact with them via stdin and stdout. Here we spawn the `ping` command with some arguments:
+In the `ping_run()` function is where the magic happens. [cockpit.spawn](https://cockpit-project.org/guide/latest/cockpit-spawn.html) lets you spawn processes on the server and interact with them via stdin and stdout. Here we spawn the `ping` command with some arguments:
 
-```javascript
-    var proc = cockpit.spawn(["ping", "-c", "4", address.val()]);
+```js
+cockpit.spawn(["ping", "-c", "4", address.val()])
 ```
 
-In a web browser you cannot block and wait until a method call completes. Anything that doesn't happen instantaneously gets its results reported back to you by [means of callback handlers](https://cockpit-project.org/guide/latest/api-cockpit.html#latest-spawn-done). jQuery has a standard interface [called a promise](http://api.jquery.com/deferred.promise/). You add handlers by calling the `.done()` or `.fail()` methods and registering callbacks. `proc.stream()` registers a callback to be invoked whenever the process produces output.
+In a web browser you cannot block and wait until a method call completes. Anything that doesn't happen instantaneously gets its results reported back to you by [means of callback handlers](https://cockpit-project.org/guide/latest/cockpit-spawn.html#cockpit-spawn-then).
+
+JavaScript has a standard interface called a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). You add handlers by calling the `.then()` or `.catch()` methods and registering callbacks (Note: historically these have been called `.done()` and `.fail()`, but these should not be used any more in new code.) The `cockpit.spawn` specific `.stream()` handler registers a callback to be invoked whenever the process produces output.
 
-```javascript
-    proc.done(ping_success);
-    proc.stream(ping_output);
-    proc.fail(ping_fail);
-    ...
+```js
+   <some promise>
+   .stream(ping_output)
+   .then(ping_success)
+   .catch(ping_fail);
 }
 ```
 
 The `ping_success()` and `ping_fail()` and `ping_output()` update the display as you would expect.
 
-So there you go ... it's a simple plugin to start off with ... next time we'll cover [how to use DBus](https://cockpit-project.org/guide/latest/api-cockpit.html#latest-dbus), and then the real fun begins.
+This should be enough for your first useful (but admittedly ugly) Cockpit page!  Please see the [Contributing](https://cockpit-project.org/external/wiki/Contributing.html) page for more links where to go from here.
```

Comment:
```suggestion
This should be enough to get you started with your first useful (but admittedly basic) Cockpit page!  Please see the [Contributing](https://cockpit-project.org/external/wiki/Contributing.html) page for more documentation on where to grow from here.
```

Is changing go to grow here too "cute"? As they're going to be growing their experience and their own custom pages, I think it's OK. But would like feedback to be sure.

Also, I changed "ugly" to "basic". It's probably better to say it this way, and should be more PC ("politically correct", not "personal computer") and friendly.